cmd/cc-uploader/main.go

```diff
@@ -189,16 +131,87 @@ func initializeServer(logger lager.Logger, uploaderConfig config.UploaderConfig,
 			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
 		}
 
-		server = &http.Server{
-			Addr:      uploaderConfig.MutualTLS.ListenAddress,
-			Handler:   ccUploaderHandler,
-			TLSConfig: clientTLSConfig,
-		}
-		return server, http_server.NewTLSServer(uploaderConfig.MutualTLS.ListenAddress, ccUploaderHandler, clientTLSConfig)
+		return http_server.NewTLSServer(uploaderConfig.MutualTLS.ListenAddress, ccUploaderHandler, clientTLSConfig)
 	}
-	server = &http.Server{
-		Addr:    uploaderConfig.ListenAddress,
-		Handler: ccUploaderHandler,
+
+	return http_server.New(uploaderConfig.ListenAddress, ccUploaderHandler)
+}
+
+func waitForDrainingToFinish() <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		uploadWaitGroup.Wait() // wait for all in-flight uploads to call Done()
+		close(done)            // signal completion
+	}()
+	return done
+}
+
+func configureServers(logger lager.Logger, uploaderConfig config.UploaderConfig, reconfigurableSink *lager.ReconfigurableSink) ifrit.Process {
+
+	var nonTLSRunner ifrit.Runner
+	tlsRunner := initializeServer(logger, uploaderConfig, true)
+	members := grouper.Members{
+		{Name: "cc-uploader-tls", Runner: tlsRunner},
+	}
+	if !uploaderConfig.DisableNonTLS {
+		nonTLSRunner = initializeServer(logger, uploaderConfig, false)
+		members = append(grouper.Members{
+			{Name: "cc-uploader", Runner: nonTLSRunner},
+		}, members...)
+	}
+	if uploaderConfig.DebugServerConfig.DebugAddress != "" {
+		members = append(grouper.Members{
+			{Name: "debug-server", Runner: debugserver.Runner(uploaderConfig.DebugServerConfig.DebugAddress, reconfigurableSink)},
+		}, members...)
+	}
+
+	group := grouper.NewOrdered(os.Interrupt, members)
+	monitor := ifrit.Invoke(sigmon.New(group))
+	logger.Info("ready")
+
+	return monitor
+}
+
+func main() {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	flag.Parse()
+
+	uploaderConfig, err := config.NewUploaderConfig(*configPath)
+	if err != nil {
+		panic(err.Error())
 	}
-	return server, http_server.New(uploaderConfig.ListenAddress, ccUploaderHandler)
+
+	logger, reconfigurableSink := lagerflags.NewFromConfig("cc-uploader", uploaderConfig.LagerConfig)
+
+	initializeDropsonde(logger, uploaderConfig)
+
+	shutdownSignal := createShutdownSignal()
+
+	monitor := configureServers(logger, uploaderConfig, reconfigurableSink)
+
+	select {
+	case err := <-monitor.Wait():
+		if err != nil {
+			logger.Info("server-exited-with-failure")
+			os.Exit(1)
+		}
+	case s := <-shutdownSignal:
+		logger.Info("shutdown-signal-received", lager.Data{"signal": s})
+
+		// Stop accepting new connections on both runners (TLS & non-TLS), Ifrit will close the listeners
+		monitor.Signal(os.Interrupt)
+
+		// Create channel to signal when uploads (including polling) are done
+		done := waitForDrainingToFinish()
+
+		select {
+		case <-done:
+			logger.Info("all-uploads-finished")
+		case <-time.After(*shutdownTimeoutInMinutes):
+			logger.Info("graceful-shutdown-timed-out",
```

Comment:
maybe warn? This is not something that should happen frequently and you may want to know. Not sure what the regular log level is for the cc-uploader, and if operators are more sensitive to warnings.

That said, they can't really do anything about it. It's an FYI.

Reply:
The logger has DEBUG, INFO, ERROR and FATAL and I think INFO is the most suiting one out of them.